### PR TITLE
[LIBCLC] Use amdgcn_fence builtin in mem_fence implementation

### PR DESCRIPTION
**Pull request by jchlanda**

### CODE REVIEW

**Comment by jchlanda** [comment]
cc: @pgorlani

File: libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl [hunk 1]
```diff
@@ -9,43 +9,45 @@
...
-    // the compiler turned any loads to scalar
-    __waitcnt(0);
-  } else if (flags & CLK_LOCAL_MEM_FENCE)
-    __waitcnt(0xff); // LGKM is [12:8]
+_CLC_DEF _CLC_OVERLOAD void __mem_fence(unsigned int scope_memory,
+                                        unsigned int semantics) {
+  switch (scope_memory) {
+  default:
+    BUILTIN_FENCE(semantics, "")
+  case 1: // Device
```
Comment (npmiller) [inline libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl:29]: We have enums for these here:

* https://github.com/intel/llvm/blob/sycl/libclc/generic/include/spirv/spirv_types.h

Comment (jchlanda) [reply libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl]: Nice, exactly what I needed.

File: libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl [hunk 2]
```diff
@@ -8,44 +8,47 @@
 
 #include <clc/clc.h>
 #include <spirv/spirv.h>
+#include <spirv/spirv_types.h>
 
-void __clc_amdgcn_s_waitcnt(unsigned flags);
+#define BUILTIN_FENCE(semantics, scope_memory)                                 \
+  if (semantics & 0x2)                                                         \
```
Comment (npmiller) [inline libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl:14]: There's also an enum in `spirv_types.h` for the semantics values

File: libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl [hunk 3]
```diff
@@ -8,44 +8,47 @@
...
-  if (flags & CLK_GLOBAL_MEM_FENCE) {
-    // scalar loads are counted with LGKM but we don't know whether
-    // the compiler turned any loads to scalar
-    __waitcnt(0);
-  } else if (flags & CLK_LOCAL_MEM_FENCE)
-    __waitcnt(0xff); // LGKM is [12:8]
+_CLC_DEF _CLC_OVERLOAD void __mem_fence(unsigned int scope_memory,
+                                        unsigned int semantics) {
+  switch ((enum Scope)scope_memory) {
+  default:
```
Comment (npmiller) [inline libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl:28]:
```suggestion
  case CrossDevice:
```

That's the last missing enum value, no need for default, is the empty string actually doing the cross device mem-fence?

Comment (jchlanda) [reply libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl:28]: Yeap, it's analogous to CUDA's `__nvvm_membar_sys` and used for an implementation of omp sytem fence: https://github.com/intel/llvm/blob/46046bd4312c3f81c300657097c480e3f56901df/openmp/libomptarget/DeviceRTL/src/Synchronization.cpp#L245

**Pull request merged**